File: Rich_Game/src/com/tw/core/toolHouse/ToolHouse.java
package com.tw.core.toolHouse;

import com.tw.core.commands.ResponsiveFactory;
import com.tw.core.house.House;
import com.tw.core.player.Player;

/**
 * Created by pzzheng on 11/12/16.
 */
public class ToolHouse extends House {
    public static final int QUIT_INDEX = -1;

    public ToolHouse(Tool... tools) {
        super(tools);
    }

    public boolean canAffordWith(int points) {
        return items.stream().anyMatch(tool -> tool.getValue() <= points);
    }

    @Override
    public Player.Status comeHere(Player player) {
        player.moveTo(this);
        if ((!canAffordWith(player.getPoints()) || player.getTools().values().stream().reduce(0, (a, b) -> a+b) == 10)) {
            return player.endTurn();
        }
        else {
            player.setResponseCommand(ResponsiveFactory.BuyTool);
            return player.waitForResponse();
        }
    }
}

File: ruby/rich_game_ruby/src/places/tool_house.rb
require_relative 'place'
require_relative '../commands/command_factory'
class ToolHouse < Place
  def initialize(*tool)
    @tools = tool
  end

  def comeHere(player)
    player.moveTo(self)
    return player.endTurn if !canAffordWith(player.asset.points)
    player.lastResponsiveCommand = CommandFactory::BuyTool
    return player.waitForResponse
  end

  def canAffordWith(points)
    @tools.each {|tool| return true if tool.value <= points}
    return false
  end


end